pkg/api/service.go

```diff
@@ -55,15 +64,63 @@ func (s *Service) Close() {
 
 func (s *Service) BatchSubscribeEnvelopes(
 	req *message_api.BatchSubscribeEnvelopesRequest,
-	server message_api.ReplicationApi_BatchSubscribeEnvelopesServer,
+	stream message_api.ReplicationApi_BatchSubscribeEnvelopesServer,
 ) error {
-	return status.Errorf(codes.Unimplemented, "method BatchSubscribeEnvelopes not implemented")
+	log := s.log.Named("subscribe")
+
+	// Send a header (any header) to fix an issue with Tonic based GRPC clients.
+	// See: https://github.com/xmtp/libxmtp/pull/58
+	err := stream.SendHeader(metadata.Pairs("subscribed", "true"))
+	if err != nil {
+		return status.Errorf(codes.Internal, "could not send header: %v", err)
+	}
+
+	requests := req.GetRequests()
+	if len(requests) == 0 {
+		return status.Errorf(codes.InvalidArgument, "missing requests")
+	}
+
+	ch, err := s.subscribeWorker.addListeners(requests)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "invalid subscription request: %v", err)
+	}
+
+	var streamLock sync.Mutex
+	for exit := false; !exit; {
+		select {
+		case envs, open := <-ch:
+			if open {
+				func() {
+					streamLock.Lock()
+					defer streamLock.Unlock()
+					err := stream.Send(&message_api.BatchSubscribeEnvelopesResponse{
+						Envelopes: envs,
+					})
+					if err != nil {
+						log.Error("sending envelope to subscribe", zap.Error(err))
+					}
+				}()
+			} else {
+				// TODO(rich) Recover from backpressure
+				log.Info("stream closed due to backpressure")
```

Comment:
With cursors, recovery from backpressure should be possible. Basically we drain the channel, and then restart the subscription from scratch using the cursor. That in turn will listen on a new channel, but pull from the DB first (ignoring duplicates coming in on the channel).

Reply:
would you explain the backpressure logic to me next time we chat?

Reply:
Will do!